raiden_contracts/tests/test_channel_unlock.py

```diff
@@ -449,3 +434,564 @@ def test_channel_settle_and_unlock(
+    assert balance_A == pre_balance_A + unlockable_A + expired_B
+    assert balance_B == pre_balance_B + unlockable_B + expired_A
+
+
+def test_unlock_twice_fails(
+        web3,
+        token_network,
+        get_accounts,
+        create_settled_channel,
+        reveal_secrets,
+):
+    (A, B) = get_accounts(2)
+    settle_timeout = 8
+
+    # Regular channel life-cycle: open -> settle -> unlock -> open -> settle -> unlock
+
+    # Mock pending transfers data
+    pending_transfers_tree_1 = get_pending_transfers_tree(web3, [1, 3, 5], [2, 4], settle_timeout)
+    reveal_secrets(A, pending_transfers_tree_1.unlockable)
+
+    # Settle the channel
+    create_settled_channel(
+        A,
+        pending_transfers_tree_1.locked_amount,
+        pending_transfers_tree_1.merkle_root,
+        B,
+        0,
+        EMPTY_MERKLE_ROOT,
+        settle_timeout,
+    )
+    token_network.functions.unlock(
+        A,
+        B,
+        pending_transfers_tree_1.packed_transfers,
+    ).transact({'from': A})
+
+    # Calling unlock twice does not work
+    with pytest.raises(TransactionFailed):
+        token_network.functions.unlock(
+            A,
+            B,
+            pending_transfers_tree_1.packed_transfers,
+        ).transact({'from': A})
+
+
+def test_channel_unlock_with_a_large_expiration(
+        web3,
+        custom_token,
+        token_network,
+        secret_registry_contract,
+        create_channel,
+        channel_deposit,
+        get_accounts,
+        close_and_update_channel,
+):
+    (A, B) = get_accounts(2)
+    settle_timeout = 8
+
+    values_A = ChannelValues(
+        deposit=20,
+        transferred=5,
+        locked=0,
+        locksroot=EMPTY_MERKLE_ROOT,
+    )
+    values_B = ChannelValues(
+        deposit=30,
+        transferred=40,
+    )
+
+    # Create channel and deposit
+    create_channel(A, B, settle_timeout)[0]
+    channel_deposit(A, values_A.deposit, B)
+    channel_deposit(B, values_B.deposit, A)
+
+    # Mock pending transfers data with large expiration date
+    pending_transfers_tree = get_pending_transfers_tree(web3, [1, 3, 5], [2, 4],
+                                                        settle_timeout + 100)
+    locksroot_bytes = get_merkle_root(pending_transfers_tree.merkle_tree)
+    values_B.locksroot = '0x' + locksroot_bytes.hex()
+    values_B.locked = get_locked_amount(pending_transfers_tree.transfers)
+
+    # Reveal secrets before settlement window ends
+    for lock in pending_transfers_tree.unlockable:
+        secret_registry_contract.functions.registerSecret(lock[3]).transact({'from': A})
+        assert secret_registry_contract.functions.getSecretRevealBlockHeight(
+            lock[2],
+        ).call() == web3.eth.blockNumber
+
+    close_and_update_channel(
+        A,
+        values_A,
+        B,
+        values_B,
+    )
+
+    # Settle channel after a "long" time
+    web3.testing.mine(settle_timeout + 1000)
```

Comment:
Original `raiden` test was:
```
""" Withdraw must accept a lock that expires after the settlement period. """
[...]
# use a really large expiration
lock_expiration = tester_chain.block.number + settle_timeout * 5

[...]
tester_chain.mine(number_of_blocks=settle_timeout + 1)
```

This test indeed uses lock `expiration > settle_timeout`.

But `web3.testing.mine(settle_timeout + 1000)` is redundant and just takes up more testing time - because you wait `1000` more blocks.
I see no reason to wait this much - pertaining to the actual test.

Reply:
We have `expiration >= settle_timeout` tested in every successfull test. I agree it's annoying to wait for it, but I'd a) remove it entirely or b) let the "large" expiration time actually pass, otherwise this is doing nothing.

Reply:
> otherwise this is doing nothing.

The test is not for testing settlement after lock expiration. The Raiden client is not aware of the expiration and will settle the first chance it gets.
This test makes sure there are no conditions that would make an unlock fail if it contains secrets with big expirations - bigger than settlement window.

Reply:
Measured mining time.
`web3.testing.mine(1000)` - `27.40565586090088` seconds
`web3.testing.mine(100)` - `0.2555418014526367` seconds
`web3.testing.mine(1)` - `0.003122091293334961` seconds

So `web3.testing.mine(100)` doesn't add that much now. It can remain like this.